From 0f3ace9f3f38e3719b83018f400d5ffa509bbd1d Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Thu, 20 Mar 2025 13:57:56 +0100
Subject: [PATCH 1/5] Make %_passwd_path and %_group_path lists

Look up multiple files if they are listes in the macros separated by
colons. This way multiple sources of user and group information can be
used.

This is needed as RPM no longer honors nsswitch as it fails for
chroots. This way one can at least configure where RPM should look for
user/group information.

Backported from commits:
eb1ee05c6d835903b3f47bc95042790e915f569e

Fixes: RHEL-78693
---
 lib/rpmug.c        | 21 ++++++++++++++++++++-
 macros.in          |  2 +-
 tests/rpmverify.at | 18 ++++++++++++++++++
 3 files changed, 39 insertions(+), 2 deletions(-)

diff --git a/lib/rpmug.c b/lib/rpmug.c
index 807677b2c..316e326f3 100644
--- a/lib/rpmug.c
+++ b/lib/rpmug.c
@@ -47,7 +47,7 @@ static const char *grpfile(void)
  * Lookup an arbitrary field based on contents of another in a ':' delimited
  * file, such as /etc/passwd or /etc/group.
  */
-static int lookup_field(const char *path, const char *val, int vcol, int rcol,
+static int lookup_field_in_file(const char *path, const char *val, int vcol, int rcol,
 			char **ret)
 {
     int rc = -1; /* assume not found */
@@ -85,6 +85,25 @@ static int lookup_field(const char *path, const char *val, int vcol, int rcol,
     return rc;
 }
 
+/*
+ * Lookup an arbitrary field based on contents of another in a ':' delimited
+ * file, such as /etc/passwd or /etc/group. Look at multiple files listed in
+ * path separated by colons
+ */
+static int lookup_field(const char *path, const char *val, int vcol, int rcol,
+			char **ret)
+{
+    ARGV_t paths = argvSplitString(path, ":", ARGV_SKIPEMPTY);
+    int rc = -1;
+    for (ARGV_t p = paths; *p; p++) {
+	rc = lookup_field_in_file(*p, val, vcol, rcol, ret);
+	if (!rc)
+	    break;
+    }
+    argvFree(paths);
+    return rc;
+}
+
 /* atol() with error handling, return 0/-1 on success/failure */
 static int stol(const char *s, long *ret)
 {
diff --git a/macros.in b/macros.in
index 7eb3d2b98..f2a28fa57 100644
--- a/macros.in
+++ b/macros.in
@@ -132,7 +132,7 @@
 
 %_keyringpath		%{_dbpath}/pubkeys/
 
-# Location of passwd(5) and group(5)
+# Location of passwd(5) and group(5), as : separated list
 %_passwd_path		/etc/passwd
 %_group_path		/etc/group
 
diff --git a/tests/rpmverify.at b/tests/rpmverify.at
index c625a1bb9..804235908 100644
--- a/tests/rpmverify.at
+++ b/tests/rpmverify.at
@@ -648,3 +648,21 @@ runroot rpm -Vv ${VERIFYOPTS} verifyfiles
 ],
 [])
 RPMTEST_CLEANUP
+
+AT_SETUP([alternative passwd location])
+AT_KEYWORDS([verify])
+RPMDB_INIT
+
+runroot rpmbuild -bb --quiet /data/SPECS/klang.spec
+runroot rpm -Uvh /build/RPMS/noarch/klang-*
+echo "klangd:x:1111:1111::/:/sbin/nologin\n" >> ${RPMTEST}/usr/lib/passwd
+echo "klangd:x:8888:" >> ${RPMTEST}/usr/lib/group
+
+RPMTEST_CHECK([
+runroot_other chown 1111:8888 /var/lib/klangd
+runroot rpm -D "_passwd_path /usr/lib/passwd:/etc/passwd" -D "_group_path /usr/lib/group:/etc/group" -V klang-server
+],
+[0],
+[],
+[])
+RPMTEST_CLEANUP
-- 
2.49.0

